src/main/java/co/cask/gcp/bigquery/sink/AbstractBigQuerySink.java

```diff
@@ -0,0 +1,419 @@
+          //timestamp for json input should be in this format yyyy-MM-dd HH:mm:ss.SSSSSS
+          json.addProperty(name, dateTimeFormatter.format(Objects.requireNonNull(input.getTimestamp(name))));
+          break;
+        default:
+          throw new IllegalStateException(
+            String.format("Field '%s' is of unsupported type '%s'", name, logicalType.getToken()));
+      }
+      return;
+    }
+
+    switch (schema.getType()) {
+      case NULL:
+        json.add(name, JsonNull.INSTANCE); // nothing much to do here.
+        break;
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+        json.addProperty(name, (Number) object);
+        break;
+      case BOOLEAN:
+        json.addProperty(name, (Boolean) object);
+        break;
+      case STRING:
+        json.addProperty(name, object.toString());
+        break;
+      default:
+        throw new IllegalStateException(
+          String.format("Field '%s' is of unsupported type '%s'", name, schema.getType()));
+    }
+  }
+
+  /**
+   * Child classes must provide configuration based on {@link AbstractBigQuerySinkConfig}.
+   *
+   * @return config instance
+   */
+  protected abstract AbstractBigQuerySinkConfig getConfig();
+
+  /**
+   * Child classes must override this method to provide specific validation logic to executed before
+   * actual {@link #prepareRun(BatchSinkContext)} method execution.
+   * For example, Batch Sink plugin can validate schema right away,
+   * Batch Multi Sink does not have information at this point to do the validation.
+   *
+   * @param context batch sink context
+   */
+  protected abstract void prepareRunValidation(BatchSinkContext context);
+
+  /**
+   * Executes main prepare run logic, i.e. prepares output for given table (for Batch Sink plugin)
+   * or for a number of tables (for Batch Multi Sink plugin).
+   *
+   * @param context batch sink context
+   * @param bucket bucket name
+   */
+  protected abstract void prepareRunInternal(BatchSinkContext context, String bucket) throws IOException;
+
+  /**
+   * Returns output format provider instance specific to the child classes that extend this class.
+   *
+   * @param configuration Hadoop configuration
+   * @param tableName table name
+   * @param tableSchema table schema
+   * @return output format provider
+   */
+  protected abstract OutputFormatProvider getOutputFormatProvider(Configuration configuration,
+                                                                  String tableName,
+                                                                  Schema tableSchema);
+
+  /**
+   * If provided dataset name does not exists in Big Query, creates it.
+   * This operation only takes place, if pipeline does not run in preview mode.
+   *
+   * @param previewEnabled indicates whether the pipeline is running in preview.
+   */
+  private void createDataset(boolean previewEnabled) throws IOException {
+    if (previewEnabled) {
```

Comment:
If the bucket does not exist then this will fail in preview? Is this intended? I think this may be confusing for the user

Reply:
Also I think this is backwards. if (previewEnabled) == true means that the pipeline is running in preview mode, where as it seems you want the opposite condition.

Reply:
Yes, my fault. We should only create bucket is we are not in preview mode. Updated the code.
In preview mode actual run is not performed thus absent dataset has no effect.